driver-core/src/main/com/mongodb/internal/connection/IdTrackingBsonWriter.java

```diff
@@ -0,0 +1,443 @@
+    private BsonValue id;
+
+    public IdTrackingBsonWriter(final BsonWriter bsonWriter) {
+        super(bsonWriter);
+        idBsonBinaryWriter = new LevelCountingBsonWriter(new BsonBinaryWriter(new BasicOutputBuffer())){};
+    }
+
+    @Override
+    public void writeStartDocument(final String name) {
+        setCurrentFieldName(name);
+
+        if (isWritingId()) {
+            idBsonBinaryWriter.writeStartDocument(name);
+        }
+        level++;
+        super.writeStartDocument(name);
+    }
+
+    @Override
+    public void writeStartDocument() {
+        if (isWritingId()) {
+            idBsonBinaryWriter.writeStartDocument();
+        }
+        level++;
+        super.writeStartDocument();
+    }
+
+    @Override
+    public void writeEndDocument() {
+        if (isWritingId()) {
+            if (idBsonBinaryWriter.getCurrentLevel() >= 0) {
+                idBsonBinaryWriter.writeEndDocument();
+            }
+            if (idBsonBinaryWriter.getCurrentLevel() == -1) {
+                if (id != null && id.isJavaScriptWithScope()) {
+                    id = new BsonJavaScriptWithScope(id.asJavaScriptWithScope().getCode(), new RawBsonDocument(getBytes()));
+                } else if (id == null) {
+                    id = new RawBsonDocument(getBytes());
+                }
+            }
+        }
+
+
+        if (level == 0 && id == null) {
+            id = new BsonObjectId();
+            writeObjectId(ID_FIELD_NAME, id.asObjectId().getValue());
+        }
+
+        level--;
+        super.writeEndDocument();
+    }
+
+    @Override
+    public void writeStartArray() {
+        if (isWritingId()) {
+            if (idBsonBinaryWriter.getCurrentLevel() == -1) {
+                idBsonBinaryWriter.writeStartDocument();
+                idBsonBinaryWriter.writeName(ID_FIELD_NAME);
+            }
+            idBsonBinaryWriter.writeStartArray();
+        }
+
+        level++;
+        super.writeStartArray();
+    }
+
+    @Override
+    public void writeStartArray(final String name) {
+        setCurrentFieldName(name);
+        if (isWritingId()) {
+            if (idBsonBinaryWriter.getCurrentLevel() == -1) {
+                idBsonBinaryWriter.writeStartDocument();
+            }
+            idBsonBinaryWriter.writeStartArray(name);
+        }
+
+        level++;
+        super.writeStartArray(name);
+    }
+
+    @Override
+    public void writeEndArray() {
+        level--;
+        super.writeEndArray();
+
+        if (isWritingId()) {
+            idBsonBinaryWriter.writeEndArray();
+            if (level == 0) {
+                idBsonBinaryWriter.writeEndDocument();
+                id = new RawBsonDocument(getBytes()).get(ID_FIELD_NAME);
+            }
+        }
+    }
+
+    @Override
+    public void writeBinaryData(final String name, final BsonBinary binary) {
+        setCurrentFieldName(name);
+        addBsonValue(() -> binary, () -> idBsonBinaryWriter.writeBinaryData(name, binary));
```

Comment:
`addBsonValue` takes two methods, one to create the BsonValue and one to write the value to the `idBsonBinaryWriter` incase we are writing a nested field.